### PR TITLE
Reuse LocalAuthentication context

### PR DESCRIPTION
**Pull request by eaceto**
### Changes

Please describe both what is changing and why this is important. Include:

- Added an instance of LAContext to the Keychain class in order to reuse it.
- Expose the LAContext as a read property in order to reuse it outside of the Keychain (In Auth0's CredentialManager, for example)

### References

Please include relevant links supporting this change such as a:

- fix issue: https://github.com/auth0/SimpleKeychain/issues/73 

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If helpful, please include manual testing steps as well. 

1. Create an instance of Keychain and configure it to use LocalAuthentication an access when the device is unlocked

```swift
keychain = A0SimpleKeychain()
keychain.useAccessControl = true
keychain.defaultAccessiblity = .whenUnlockedThisDeviceOnly
keychain.setTouchIDAuthenticationAllowableReuseDuration(5.0)

let r1 = keychain.setString("TEST 1A", forKey: "KEY 1", promptMessage: "P1")
let r2 = keychain.setString("TEST 2A", forKey: "KEY 2", promptMessage: "P2")       

let v1 = keychain.string(forKey: "KEY 1", promptMessage: "P1B")
let v2 = keychain.string(forKey: "KEY 2", promptMessage: "P2B")
```

2. Enable FaceID via declaration in Info.plist

```
<key>NSFaceIDUsageDescription</key>
<string>Access Keychain</string>
```

3. When running the code in (1) the Passcode / Touch ID / Face ID local authentication will be asked only once (will be reused). With the current code it's asked every time a keychain operation is performed.

[ ] This change adds unit test coverage (or why not)

It does not include Unit Test. The Test requires UI Interaction in order to see the behavior

[X] This change has been tested on the latest version of the platform/language or why not

Tested on iOS 13.2

### Checklist

[X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)

[X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

[X] All existing and new tests complete without errors

### CODE REVIEW

**Comment by Widcket** [comment]
Hi @eaceto, apologies for the delay. `LAContext` [is only available on iOS and macOS](https://developer.apple.com/documentation/localauthentication/lacontext?language=objc). This library supports tvOS and watchOS as well, that's why the CI build is failing. Can you please add platform checks to the code in order to maintain compatibility?

**Comment by eaceto** [comment]
Hello Rita, how are you? Thanks for replying, and sorry for not fixing that earlier. I was on holidays.

Let me add this checks now in order to fix the CI errors. Thanks!

**Comment by eaceto** [comment]
Hi @Widcket ! All tests passed, now. If you want to perform a detailed review on the Pull Request, please let me know.

File: SimpleKeychain/A0SimpleKeychain.m [hunk 1]
```diff
@@ -44,10 +44,33 @@ - (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)acce
...
+        }
+#endif
     }
     return self;
 }
 
+- (void)setTouchIDAuthenticationAllowableReuseDuration:(NSTimeInterval) duration {
+// // This does not apply to watchOS & tvOS
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
+    if (@available(iOS 8, macOS 10.12, *)) {
```
Comment (Widcket) [inline SimpleKeychain/A0SimpleKeychain.m]: Why the double check?

File: SimpleKeychain/A0SimpleKeychain.m [hunk 2]
```diff
@@ -44,10 +44,33 @@ - (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)acce
...
+            _localAuthenticationContext = [LAContext new];
+            _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = 0;
+        }
+#endif
     }
     return self;
 }
 
+- (void)setTouchIDAuthenticationAllowableReuseDuration:(NSTimeInterval) duration {
+// // This does not apply to watchOS & tvOS
```
Comment (Widcket) [inline SimpleKeychain/A0SimpleKeychain.m]: Double pair of slashes.

File: SimpleKeychain/A0SimpleKeychain.m [hunk 3]
```diff
@@ -44,10 +44,33 @@ - (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)acce
         _accessGroup = accessGroup;
         _defaultAccessiblity = A0SimpleKeychainItemAccessibleAfterFirstUnlock;
         _useAccessControl = NO;
+        
+// This does not apply to watchOS & tvOS
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
+        if (@available(iOS 8, macOS 10.12, *)) {
```
Comment (Widcket) [inline SimpleKeychain/A0SimpleKeychain.m]: Why the double check?

File: SimpleKeychain/A0SimpleKeychain.m [hunk 4]
```diff
@@ -279,7 +302,15 @@ - (NSMutableDictionary *)baseQuery {
     NSMutableDictionary *attributes = [@{
                                          (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
                                          (__bridge id)kSecAttrService: self.service,
+                                         (__bridge id)kSecUseAuthenticationUI: (__bridge id)kSecUseAuthenticationUIAllow,
                                          } mutableCopy];
+    
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
+    if (@available(iOS 8, macOS 10.12, *)) {
```
Comment (Widcket) [inline SimpleKeychain/A0SimpleKeychain.m]: Why the double check?

**Comment by eaceto** [comment]
@Widcket all changes where made, and branch was rebased to current master (due to a PR about SPM)

File: SimpleKeychain/A0SimpleKeychain.h [hunk 5]
```diff
@@ -22,6 +22,10 @@
 
 #import <Foundation/Foundation.h>
 
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.h]: A simpler check is to see if the header is available 

```objc
#if __has_include("LocalAuthentication/LocalAuthentication.h")
```

File: SimpleKeychain/A0SimpleKeychain.h [hunk 6]
```diff
@@ -141,6 +145,14 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (assign, nonatomic) BOOL useAccessControl;
 
+
+/**
+*  LocalAuthenticationContext used to access items. Default value is a new LAContext object
+*/
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.h]: Not sure on the logic here to also check for the library ! to be a specific version, would be better I think to use MIN and you only need to specify for iOS and Mac OS e.g.

```
#if __IPHONE_OS_VERSION_MIN_ALLOWED >= 90000
```

File: SimpleKeychain/A0SimpleKeychain.h [hunk 7]
```diff
@@ -171,6 +183,14 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (instancetype)initWithService:(NSString *)service accessGroup:(nullable NSString *)accessGroup;
 
+/**
+ *  The duration for which Touch ID authentication reuse is allowable.
+ *  Maximun value is LATouchIDAuthenticationMaximumAllowableReuseDuration
+ */
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.h]: Same as before

File: SimpleKeychain/A0SimpleKeychain.m [hunk 8]
```diff
@@ -44,10 +44,29 @@ - (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)acce
         _accessGroup = accessGroup;
         _defaultAccessiblity = A0SimpleKeychainItemAccessibleAfterFirstUnlock;
         _useAccessControl = NO;
+        
+// This does not apply to watchOS & tvOS
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.m]: same as before

File: SimpleKeychain/A0SimpleKeychain.m [hunk 9]
```diff
@@ -44,10 +44,29 @@ - (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)acce
...
+        _localAuthenticationContext = [LAContext new];
+        _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = 0;
+#endif
     }
     return self;
 }
 
+- (void)setTouchIDAuthenticationAllowableReuseDuration:(NSTimeInterval) duration {
+// This does not apply to watchOS & tvOS
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.m]: same as before

File: SimpleKeychain/A0SimpleKeychain.m [hunk 10]
```diff
@@ -279,7 +298,13 @@ - (NSMutableDictionary *)baseQuery {
     NSMutableDictionary *attributes = [@{
                                          (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
                                          (__bridge id)kSecAttrService: self.service,
+                                         (__bridge id)kSecUseAuthenticationUI: (__bridge id)kSecUseAuthenticationUIAllow,
                                          } mutableCopy];
+    
+#if (defined __IPHONE_8_0 || defined __MAC_10_12) && !defined __TVOS_9_0 && !defined __WATCHOS_1_0
```
Comment (cocojoe) [inline SimpleKeychain/A0SimpleKeychain.m]: same as before

File: SimpleKeychain/A0SimpleKeychain.h (hunk 5 above)
Comment (eaceto) [reply SimpleKeychain/A0SimpleKeychain.h]: Great tip! I didn't know about that macro.

File: SimpleKeychain/A0SimpleKeychain.h (hunk 6 above)
Comment (eaceto) [reply SimpleKeychain/A0SimpleKeychain.h]: You are right. Would you agree to create a Macro in A0SimpleKeychain.h with the following definition?

```
#define A0LocalAuthenticationCapable __IPHONE_OS_VERSION_MIN_ALLOWED >= 80000 || __MAC_OS_X_VERSION_MIN_REQUIRED >= 101200
```
So when necessary, we can use it like this

```
#if A0LocalAuthenticationCapable
   ...
#endif
```

Thanks!

**Pull request merged**